### ISSUE TITLE
Failing to parse TypeScript files outside the src directory

### ISSUE THREAD

**Issue by FredyC** [storybookjs/storybook#11644]
**Describe the bug**
I am utilizing TypeScript path aliases with CRA along with Yarn workspaces in a mono repo situation.

Importing files from the aliased path that contains type declaration fails to parse that file when compiling Storybook. The CRA itself has no trouble with that.

Upon removing type declaration the import is working correctly.

**To Reproduce**
I did set up a whole repo to reproduce a problem. Instructions are pretty straightforward, but I've outlined in README the most notable differences from the regular setup.

https://github.com/FredyC/storybook-cra-workspaces

**Expected behavior**
Files with Typescript declaration should be imported without an issue.

**Screenshots**
![image](https://user-images.githubusercontent.com/1096340/88175563-5085f000-cc26-11ea-9ca0-744b22c552b8.png)

**System:**
![image](https://user-images.githubusercontent.com/1096340/88175689-80cd8e80-cc26-11ea-95e9-bd6f664df7ae.png)

**Comment by shilman** [comment]
cc @mrmckeb

**Comment by mrmckeb** [comment]
Hi @FredyC, this preset aligns with how CRA handles files - which is on purpose, so that users don't get a false sense of security when writing stories. We want anything developed in Storybook to work with the app.

The rules around not handling files outside of `src` come from the CRA Webpack config, and the only change we make is to support the Storybook config folder.

Now, if you're using a monorepo and build each package, this will work. We do this with our projects in my company and it works really well, CRA and Storybook will even pick up those changes in real time.

So if you have some common tools, you could place those in a monorepo package (maybe called `common`) and then just import as `common` in your project (without relative paths). Again, we do this at my company, and I do this on side projects too.

---

If you want to make this work, you can modify the Webpack config by following these instructions:
https://storybook.js.org/docs/configurations/custom-webpack-config/

You can see some examples of how we added support for the config directory here:
https://github.com/storybookjs/presets/search?q=configDir&unscoped_q=configDir

Most of the work happens here:
https://github.com/storybookjs/presets/blob/master/packages/preset-create-react-app/src/helpers/processCraConfig.ts

Let me know if this answers your question or not, thanks!

**Comment by FredyC** [comment]
@mrmckeb It's not true what you are saying about CRA. If you try my repo, you can run it just fine. I think that core magic comes from Typescript path aliases which makes it look like it's coming from `src`.

Also as I said, file outside of `src` can be imported within Storybook as long as there are no types in that files. That would imply that there some misconfiguration in loaders perhaps?

I don't want to explicitly rely on `common` because it would mean I would have rebuild every time I change file in any of those _packages_. CRA or Storybook is unable to see those changes. That would horribly hinder the DX. With my solution, I can do the change and everything works right away.

**Comment by mrmckeb** [comment]
@FredyC, I see that you've updated the comment - but what I'm saying is definitely true. I'm one of the maintainers of CRA, alongside Storybook, and the preset for CRA.

I see that you're not actually using `react-scripts`, you're using `react-app-rewired` - and you have a plugin to enable aliasing.
https://github.com/FredyC/storybook-cra-workspaces/blob/master/modules/web/config-overrides.js#L7

As I mentioned, you can definitely make this work on your own by modifying the Webpack config in Storybook's `main.js` (or `main.ts`) file as I mentioned above. If that doesn't work for you - or you have any trouble getting that configured - please let me know.

**Comment by FredyC** [comment]
@mrmckeb Thank you for your patience. Have you also noticed, I am using the same aliasing config in Storybook webpack config?

https://github.com/FredyC/storybook-cra-workspaces/blob/master/modules/web/.storybook/main.js#L4

There is no other special webpack configuration done for the CRA. That's why I am not sure why Storybook is not working. There is no problem in finding those aliased imports, but transpiling Typescript files (with explicit declaration) is problematic for some reason.

**Comment by mrmckeb** [comment]
Yes, I saw that - which I might have thought would work too, but can see it doesn't.

The problem is that code assumes a pure CRA Webpack config. However, the CRA preset takes the CRA config, modifies it a little (so that it works in the Storybook environment) and then returns it. So it's likely that your issue is being caused by these differences.

There is a way you can log out your Webpack config in Storybook which might help to understand where the issue lies:
https://storybook.js.org/docs/configurations/custom-webpack-config/#debug-the-default-webpack-config

You should then be able to see the issue (hopefully it's easy to spot). I took a look at the plugin you're using (`react-app-rewire-alias`), and I didn't see anything obvious... I'm a little busy today/tomorrow, but if you can't see anything after debugging, let me know and I'll dig through your code and help get this working :)

**Comment by FredyC** [comment]
Well, the only modification I am doing is adding those webpack aliases to the original CRA config. There is no hidden magic or webpack plugin behind it. When I do the same in Storybook config it's pretty safe to assume it will behave the same way.

I think there is something in Storybook webpack config that's different from CRA's one that prevents this approach. I tried to logging out the config, but it spilled way too much information for me to be able to find something in there.

**Comment by FredyC** [comment]
@mrmckeb Can you please reopen, this is still a valid issue. I stopped using Storybook for now because of this issue, it was too limiting not to be able to use TS, but I would certainly like to solve it.

**Comment by Deschain23** [comment]
I'm having the same error. From me is throwing it when declaring a type:
![image](https://user-images.githubusercontent.com/31232086/97227044-87761f80-179a-11eb-8a81-6068f5e13294.png)

Any update on how to resolve this issue?

**Comment by FredyC** [comment]
Yep, still the problem. Can someone pin this so it doesn't get inactive?

**Comment by sjuchno** [comment]
I added a small modification to @Deschain23 solution so you don't need to worry about proper indexes
```
  webpackFinal: webpackConfig => {
    const scopePluginIndex = webpackConfig.resolve.plugins.findIndex(
      ({ constructor }) => constructor && constructor.name === 'ModuleScopePlugin'
    );

    webpackConfig.resolve.plugins.splice(scopePluginIndex, 1);

    // https://github.com/storybookjs/storybook/issues/11644#issuecomment-716815342
    const oneOfIndex = webpackConfig.module.rules.findIndex(rule => !!rule.oneOf);
    const tsIndex = webpackConfig.module.rules[oneOfIndex].oneOf.findIndex(item => item.test.test('.tsx'));
    webpackConfig.module.rules[oneOfIndex].oneOf[tsIndex].include.push(path.resolve(__dirname, '../../shared/src'));

    return webpackConfig;
  },
```

**Comment by sparky-raccoon** [comment]
Came across this issue. Depending on your storybook / cra version, the solution above might be outdated, because rules are added / moved and indexes are no longer relevant. I've written this to ensure that no matter the initial webpack configuration, the external directory path will be included where it should be. It's quite naive but it did the trick for me.

```
webpackFinal: config => {
    const destinationPath = path.resolve(__dirname, '../../../package');
    const addExternalPath = rules => {
      for (let i = 0; i < rules.length; i++) {
        const rule = rules[i];
        if (rule.test && RegExp(rule.test).test('.tsx')) {
          if (rule.include?.length) rule.include.push(destinationPath);
          else rule.include = destinationPath;
        } else if (rule.oneOf) {
          addExternalPath(rule.oneOf);
        }
      }
    };

    addExternalPath(config.module.rules);

  return config;
},

**Comment by valentinpalkovic** [comment]
Closing as obsolete. Create React App is deprecated upstream, and Storybook's recommended React paths are now Vite, Next.js, or RSBuild. Monorepo TS aliases across yarn workspaces are handled natively by those modern builders. If you still need TS aliases with CRA on Storybook 10, please open a new issue against `@storybook/preset-create-react-app` with a current repro.
